stmhal/sdcard.c

```diff
@@ -85,6 +86,29 @@ void sdcard_init(void) {
     // invalidate the sd_handle
     sd_handle.Instance = NULL;
 
+    // configure the SD card detect pin
+    // we do this here so we can detect if the SD card is inserted before powering it on
+    GPIO_Init_Structure.Mode = GPIO_MODE_INPUT;
+    GPIO_Init_Structure.Speed = GPIO_SPEED_LOW;
+#if defined(MICROPY_HW_SDCARD_DETECT_PIN)
+    // assume a mechanical switch is used
+    GPIO_Init_Structure.Pull = MICROPY_HW_SDCARD_DETECT_PULL;
+    GPIO_Init_Structure.Pin = MICROPY_HW_SDCARD_DETECT_PIN.pin_mask;
+    HAL_GPIO_Init(MICROPY_HW_SDCARD_DETECT_PIN.gpio, &GPIO_Init_Structure);
+#elif defined(MICROPY_HW_SDCARD_DETECT_PULL)
+    // assume no mechanical switch, so assume CD/DAT3 is used for card detect.
+    GPIO_Init_Structure.Pull = MICROPY_HW_SDCARD_DETECT_PULL;
+    GPIO_Init_Structure.Pin = pin_C11.pin_mask;
+    HAL_GPIO_Init(pin_C11.gpio, &GPIO_Init_Structure);
+    // capture sdcard present state before the IO pin is remapped to SDIO functionality
+    sdcard_present = (HAL_GPIO_ReadPin(pin_C11.gpio, pin_C11.pin_mask) == MICROPY_HW_SDCARD_DETECT_PRESENT);
```

Comment:
Ideally yes, but that can only be done if there is an independent signal to read whether there is an SD card present.  It could be done with CD/DAT3 if using SPI mode as only a single pin is used for data transfer.  I'm unsure if SD card detection can be achieved with CD/DAT3 if using SDIO mode.  My assumption is/was "no", but I don't really know.

I guess it could be done by sending a command and monitoring the response, but I'm not sure how efficient that would be.